viamrobotics/slam
# RSDK-1075 [carto] Add GetInternalState

This PR allows for the on-demand serialization of the internal state (.pbstream) of cartographer. This can then be saved to a file and loaded into a subsequent cartographer run. Currently addition PRs are in the work to do this for OrbSLAM, adding the GetInternalState endpoint to RDK and including GetInternalState tests to the integration testing scripts.


JIRA Ticket: [RSDK-1075](https://viam.atlassian.net/browse/RSDK-1075)

RDK PR/ Cartographer Testing (adding GetInternalState): [PR #1776](https://github.com/viamrobotics/rdk/pull/1776)


Notes:
The implementation in this PR utilizes a file interface to access the data and sent it over grpc to RDK. This is no ideal as it contains a file read and write. However, this is hard to get away from which I found after about 6 hours of research. 

Cartographer has two functions for serializing the state `SerializeState` and `SerializeStateToFile`. However this is no entirely true as `SerializeState` performs the exact same function as `SerializeStateToFile` except it requires the user to define  a `ProtoStreamWriter`. normally you'd think that we could then use this data object to access the serialized data and prevent it being added to a file. 

However, due to the strict type definition of `ProtoStreamWriter`, we not only cannot access the data directly, but it is stored in an `ofstream` which means it will write any data added to it directly to the filename set upon its definition. The only possible solutions to this is to either:

- Attempt to copy over half of `mapping/io/internal` along with `proto_stream.h/.cc` and `proto_stream_interface.h` to allow a viam version to be created and then relink these files. (2-3d)
- Edit cartographer itself which we have avoided doing up until now (1d)
- Manually serialize each part of state which will involves lots of copying and relinked (1-2d).

For now, I've gone with the much simpler file interface method and deleteing the produced temp file. If we wish to pursue others we should discuss whether that be done now or later. Additionally, if anyone wants to run though the logic with me feel free to PM me directly. 

Testing:
This PR was tested on both the integration test data set as well as the large 5MB office dataset without issues. To replicate said testing you can follow the instructions below:

1. Install grpcurl (go install github.com/fullstorydev/grpcurl/cmd/grpcurl@latest) and jq (sudo apt-get install jq) 

2. Start RDK server with either live data or an appropriate offline dataset

3. Run agrpcurl current request of the GetInternalState endpoint `grpcurl -max-msg-sz 10000000 -d '{"name": "slam"}' -plaintext -protoset <(~/slam/slam-libraries/grpc/bin/buf build -o -) $(sudo netstat -tulpn | grep LISTEN | grep carto_grpc_ | awk '{print $4}' | sort | head -n 1) viam.service.slam.v1.SLAMService/GetInternalState >> ~/output_raw.pbstream`
Note: You may need to edit the grep part of the command due to it being cut off

4. Transform the output into base64 using jq `cat ~/output_raw.pbstream | jq -r ".internalState" | base64 -d >> ~/output.pbstream`

5. Move `~/output.pbstream` into your data folder and rename it to a valid map name/timestamp (Note: if you are running on offline data you should also make sure the timestamp for this file is before the last of your data files)

6. Re run RDK server and confirm that cartographer starts processing data without issue. If an issue occurs it will cause constant restarts of the binary



[RSDK-1075]: https://viam.atlassian.net/browse/RSDK-1075?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
